spec/datadog/profiling/stack_recorder_spec.rb

```diff
@@ -39,6 +39,41 @@ def slot_two_mutex_locked?
     end
   end
 
+  shared_examples_for 'locking behavior' do |operation|
+    context 'when slot one was the active slot' do
+      it 'sets slot two as the active slot' do
+        expect { stack_recorder.public_send(operation) }.to change { active_slot }.from(1).to(2)
+      end
+
+      it 'locks the slot one mutex and keeps it locked' do
```

Comment:
Isn't the part "and keeps it locked" redundant?
```suggestion
      it 'locks the slot one mutex' do
```

Reply:
Actually a better question.
Aren't these 3 conditions `change { active_slot }.from(1).to(2)`, `change { slot_one_mutex_locked? }.from(false).to(true)`, and `change { slot_two_mutex_locked? }.from(true).to(false)` actually representing one single thing: "the slot was flipped"?

Reply:
> Isn't the part "and keeps it locked" redundant?

Good point! Will simplify.

> Actually a better question.
> Aren't these 3 conditions change { active_slot }.from(1).to(2), change { slot_one_mutex_locked? }.from(false).to(true), and change { slot_two_mutex_locked? }.from(true).to(false) actually representing one single thing: "the slot was flipped"?

Yes; "slot being flipped" means 3 different observable effects:

* `state->active_slot` gets changed
* `slot_one_mutex` gets locked/unlocked
* `slot_two_mutext` gets locked/unlocked

I chose to test the three effects here in a more "white-box way", but indeed this wasn't the only option. For instance, I could add a RSpec helper to represent the three checks, or even move that to the `StackRecorder` class itself (using Ruby or C code).

This spec is effectively asserting on private state of the class, rather than the class abstracting this away and the spec asserting on the higher-leve  abstraction of flipping. BUT, since the class itself is already quite complex (especially being in C), I thought having more of it documented and tested on the Ruby side makes it more understandable, and so worthy of the exception to the usual "don't test private stuff".

Reply:
Descriptions improved in 1f2fdf5b8d45fb43dfbd1fff21a66923036a6255